# Differential-Expression-Analysis
How to perform differential expression analysis on RNA-Seq data

## Input Data:
- Raw Reads from sequencing, typically in fastq or fastq.gz format
- Reference genome in fasta format and GFF file (https://en.wikipedia.org/wiki/General_feature_format)
  - Ex: https://www.ncbi.nlm.nih.gov/assembly/GCF_000002985.6/

## Workflow:

### 1. Trim raw reads with Trimmomatic (http://www.usadellab.org/cms/?page=trimmomatic): 

```
mkdir trimmed-reads

trimmomatic PE -threads 24 $forward $reverse\ 
    trimmed-reads/$paired_forward trimmed-reads/$unpaired_forward\
    trimmed-reads/$paired_reverse trimmed-reads/$unpaired_reverse\
    ILLUMINACLIP:$ADAPTERS:2:30:10:8:true\
    LEADING:3 TRAILING:3\
    SLIDINGWINDOW:4:10 MINLEN:36
```
This will perform trimming on one set of paired end reads.
Replace `$forward` with the R1 reads and `$reverse` with the R2 reads for each sample.
Replace `$paired_forward`, `$unpaired_forward`, `$paired_reverse`, and `$unpaired_reverse` with the desired names.
`$ADAPTERS` must be a path to a fasta file with the sequencing adapters used.


### 2. Build index for reference genome with hisat2-build (https://daehwankimlab.github.io/hisat2/manual/)

```
hisat2-build <reference_in> <ht2_base>

hisat2-build GCF_000002985.6_WBcel235_genomic.fna GCF_000002985.6_WBcel235_genomic.fna.idx
```

This will create several index files for your reference.
```
GCF_000002985.6_WBcel235_genomic.fna.idx.1.ht2
GCF_000002985.6_WBcel235_genomic.fna.idx.2.ht2
...
GCF_000002985.6_WBcel235_genomic.fna.idx.8.ht2
```

### 3. Map paired and trimmed reads to reference genome using hisat2
```
hisat2 [options]* -x <hisat2-idx> {-1 <m1> -2 <m2> | -U <r> | --sra-acc <SRA accession number>} [-S <hit>]

hisat2 -x GCF_000002985.6_WBcel235_genomic.fna.idx -1 $paired_forward -2 $paired_reverse -S $sam_name -p 24 --max-intronlen 10000
```
This will map one set of paired and trimmed reads to the reference genome.
Replace `$paired_forward` with the trimmed paired R1 reads and `$paired_reverse` with the trimmed paired R2 reads for each sample.
`$sam_name` refers to the output name of the alignment file, this file will be in the SAM format (https://samtools.github.io/hts-specs/SAMv1.pdf)

### 4. Counting reads in features with htseq-count
```
htseq-count [options] <alignment_files> <gff_file>

htseq-count -f sam $sam_name GCF_000002985.6_WBcel235_genomic.gff -t gene > $sam_name.features.txt
```
This will count the reads for one sample.
Replace `$sam_name` with the SAM file for each sample.

### 5. Build count matrix using bash commands
The output `sample.sam.features.txt` will be a table with two columns. The first column will be gene names, and the second will be respective counts for the sample.
To combine these individual counts files into a single matrix we can use some bash commands.
```
# pick any one sample features table for the first step
echo "" > genes.mat
cat sample1.sam.features.txt | cut -f 1 >> genes.mat

# loop through sample feature tables to make individual .mat files
for table in *features.txt
do
  sample_name=$(basename $table | cut -f 1 -d'.')
  echo $sample_name >> $sample_name.mat
  cat $table | cut -f 2 >> $sample_name.mat
done

# paste tables together into single matrix
paste -d'\t' genes.mat sample1.mat ... sampleX.mat > gene.counts.matrix
```
The resulting file gene.counts.matrix can be used in DESeq2. It should have one column with gene names, and one column per sample with counts.

View this file by typing `less -S gene.counts.matrix`

### 6. Running DESeq2 (Python Implementation)
The script `unfiltered-deseq.py` can be used to run DESeq2 in Python. This requires the pydeseq2 library from https://github.com/owkin/PyDESeq2.
For visualization, bioinfokit is used (https://github.com/reneshbedre/bioinfokit)
Both libraries can be installed using pip.

Once all libraries are installed, test the command using `./unfiltered-deseq.py -h`. This will bring up the help menu.
```
usage: unfiltered-deseq.py [-h] [--alpha ALPHA] [--no_cooks] [--no_independent] [--plot] [--save] [--abnormal ABNORMAL [ABNORMAL ...]] [--normal NORMAL [NORMAL ...]]
                           [--out OUT]
                           matrix

Performs DESeq2 in python

positional arguments:
  matrix                Read counts matrix generated by kallisto

options:
  -h, --help            show this help message and exit
  --alpha ALPHA, -p ALPHA
                        P-value and adjusted p-value significance threshold
  --no_cooks            whether to not filter and refit cooks outliers
  --no_independent      whether to not independent filter
  --plot                whether to plot final data
  --save                whether to store intermediates in pkl format
  --abnormal ABNORMAL [ABNORMAL ...]
                        sample prefix for abnormal samples
  --normal NORMAL [NORMAL ...]
                        sample prefix for normal samples
  --out OUT, -o OUT     output prefix for all files
```
Different options can be changed depending on desired alpha value, and whether cooks refitting, independent filtering, and plotting data are desired.
In addition you must specify the abnormal and normal sample prefixes.

If we want to run DESeq2 with alpha of 0.05, no cooks refitting, samples 1-3 as normal and samples 4-6 as abnormal, we can run the following:
```
./unfiltered-deseq.py gene.counts.matrix --alpha 0.05 --no_cooks --plot --normal sample1 sample2 sample3 --abnormal sample4 sample5 sample6 --out deseq
```
